### PR TITLE
fix(github): bump upload/download artifact actions to v4; resolves runtime deprecation warnings

### PR DESCRIPTION
**Pull request by icj217**
Update artifact-related github actions that are generating warnings in projects due to nodejs 16 being deprecated in github actions runners.

v4 of actions/upload-artifact [introduced](https://github.com/actions/upload-artifact/blob/main/docs/MIGRATION.md) some breaking changes, so we've set `overwrite` to be true by default to mirror the previous behavior as closely as possible.

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by mrgrain — CHANGES_REQUESTED** [review]
Can you keep this limited to upload/download and split other actions in to separate PRs please?

**Comment by icj217** [comment]
> Can you keep this limited to upload/download and split other actions in to separate PRs please?

Sure thing. [Here's the separate PR](https://github.com/projen/projen/pull/3338)

File: src/github/workflow-steps.ts [hunk 1]
```diff
@@ -224,6 +225,15 @@ export interface UploadArtifactWith {
    * @default 6
    */
   readonly compressionLevel?: number;
+
+  /**
+   * Whether action should overwrite an existing artifact with the same name (should one exist)
+   *
+   * Introduced in v4 and represents a breaking change from the behavior of the v3 action.
+   *
+   * @default true (to maintain backwards compatibility)
```
Comment (mrgrain) [inline src/github/workflow-steps.ts:234]:
```suggestion
   * Introduced in v4 and represents a breaking change from the behavior of the v3 action.
   * To maintain backwards compatibility with existing, this should be set the `true` (the default).
   *
   * @default true
```

**Review by mrgrain — APPROVED** [review]
:shipit:

**Pull request merged**